EIPS/eip-6963.md

```diff
@@ -0,0 +1,118 @@
+---
+eip: 6963
+title: Multi Injected Provider Interface (aka MIPI)
+description: Using `window.evmproviders` instead of `window.ethereum`
+author: Pedro Gomes (@pedrouid), Kosala Hemachandra (@kvhnuke), Richard Moore (@ricmoo), Gregory Markou (@GregTheGreek)
+discussions-to: https://ethereum-magicians.org/t/eip-6963-multi-injected-provider-interface-aka-mipi/14076/2
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-05-01
+requires: 1193
+---
+
+## Abstract
+
+A Javascript interface injected by browser extensions installed by the user that enables user selection from multiple wallet providers using a window object labelled `window.evemproviders` instead of overwriting `window.ethereum` object.
+
+## Motivation
+
+Currently, wallet providers that offer browser extensions must inject their Ethereum providers (EIP-1193) into the same window object `window.ethereum` however this creates conflicts for users that may install more than one browser extension.
+
+Browser extensions are loaded in the web page and do not have a predictable pattern resulting in a race condition where the user is not in control to choose which wallet provider must be selected for exposing Ehtereum interface under the `window.ethereum` object.
+
+This results not only in a degraded user experience but also increases the barrier to entry for new browser extensions as users are forced to only install one browser extension at a time.
+
+Some browser extensions attempt to counteract this problem by delaying their injection to overwrite the same `window.ethereum` object which creates an unfair competition for wallet providers and lack of interoperability.
+
+In this proposal, we provide a solution that optimitizes for interoperability and enables fairer competition by lowering the barrier to entry for new wallet providers to improve the user experience on Ethereum.
+
+This is achieved by replacing the current window object `window.ethereum` with a new window object `window.evmproviders` that enables multiple browser extensions to be exposed for user selection in the web page.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Provider Info
+
+Each wallet provider will be exposed with the following interface `ProviderInfo` that will be used to display to the user:
+
+```typescript
+/**
+ * Represents the assets needed to display a wallet
+ */
+interface ProviderInfo {
+  uuid: string;
+  walletId: string;
+  name: string;
+  icon: string;
+}
+```
+
+The values in the `ProviderInfo` MUST be used as follows:
+
+- uuid - locally unique of the wallet provider (UUIDv4 compliant)
```

Comment:
I was also going to ask the same question.

Reply:
This was feedback from the TG group that we had created for EIP-5749

The feedback was that `uuid` were insufficient to identify wallets thus we needed `walletId` 

However using just `walletId` would not allow us to prevent duplicate or wallet providers overriding each other

Having two identifiers prevents these conflicts

Reply:
I missed that earlier conversation - is this `uuid` a permanently fixed value or set on page load? i.e. is the guidance here that a wallet extension should ship with a uuid that won't change across versions?

Reply:
How does a local id prevent conflicts? How would you distinguish between two arbitrary local ids? 

Expanding on @ahollenbach's question, if the `uuid` is a deterministic fixed value, how is that different from the `walletId`? Another extension could still disguise themselves as that `uuid` if it is fixed. But, if the `uuid` is non-deterministic & set on page load, how do we distinguish between them (as I mentioned above)?

Feel like a global `walletId` is enough.

Reply:
I'm not really understanding how both a `uuid` and `walletId` prevents conflicts.

Expanding on @ahollenbach's question, if the `uuid` is a deterministic fixed value, how is that different from the `walletId`? Another extension could still disguise themselves as that `uuid` if so; which means that `walletId` and `uuid` **are not** mutually exclusive so a global `walletId` is probably enough. 

On the other hand, if the `uuid` is arbitrary/non-deterministic & set on page load, how do we distinguish between them (as I mentioned above)?